rse-bash.Rmd

```diff
@@ -1256,801 +1176,895 @@ $ ls lengths.txt
-of the original files, naming the copies `original-basilisk.dat` and `original-unicorn.dat`.
-We can't use:
+-   `!head` re-runs the most recent command starting with `head`,
+    while `!wc` re-runs the most recent starting with `wc`.
+-   If we type <kbd>Ctrl</kbd>+<kbd>R</kbd>
+    the shell does an interactive reverse search for whatever we type next.
+    If we don't like the first thing it finds,
+    we can type <kbd>Ctrl</kbd>+<kbd>R</kbd> again to go further back.
+
+If we use `history`, <kbd>↑</kbd>, or <kbd>Ctrl</kbd>+<kbd>R</kbd>,
+we will quickly notice that loops don't have to be broken across lines.
+Instead,
+their parts can be separated with semi-colons:
 
 ```shell
-$ cp *.dat original-*.dat
+$ for filename in *_temp.csv; do head -n 4 $filename | tail -n 3; done
 ```
 
-because that would expand to:
+This is fairly readable,
+although even experienced users have a tendency to put the semi-colon after `do` instead of before it.
+If our loop contains multiple commands,
+though,
+the multi-line format is much easier to read—compare this:
 
 ```shell
-$ cp basilisk.dat unicorn.dat original-*.dat
+$ for filename in *_temp.csv
+> do
+>   echo $filename
+>   head -n 4 $filename | tail -n 3
+> done
 ```
 
-This wouldn't back up our files, instead we get an error:
+with this:
 
-```text
-cp: target `original-*.dat' is not a directory
+```shell
+$ for filename in *_temp.csv; do echo $filename; head -n 4 $filename | tail -n 3; done
 ```
 
-This problem arises when `cp` receives more than two inputs. When this happens, it
-expects the last input to be a directory where it can copy all the files it was passed.
-Since there is no directory named `original-*.dat` in the `creatures` directory we get an
-error.
+## How can I create new filenames automatically? {#rse-bash-autoname}
 
-Instead, we can use a loop:
+Suppose we want to create a backup copy of each precipitation data file.
+If we don't want to change the files' names,
+we can do this with `cp`:
 
 ```shell
-$ for filename in *.dat
-> do
->     cp $filename original-$filename
-> done
+$ cd ~/climate-data
+$ mkdir backup
+$ cp cleaned/*_prec.csv backup
 ```
 
-This loop runs the `cp` command once for each filename.
-The first time,
-when `$filename` expands to `basilisk.dat`,
-the shell executes:
+But what if we want the backup files to have the [extension][filename-extension] `.bak` instead of `.csv`?
+`cp` can do this for a single file:
 
 ```shell
-cp basilisk.dat original-basilisk.dat
+$ cp cleaned/andamooka_prec.csv backup/andamooka_prec.bak
 ```
 
-The second time, the command is:
+but can't do all the files at once:
 
 ```shell
-cp unicorn.dat original-unicorn.dat
+$ cp cleaned/*_prec.csv backup/*_prec.bak
 ```
 
-Before running a loop,
-it can be a good idea to check that the loop is doing the correct thing.
-We learned earlier how to print strings using `echo`,
-and we can modify the loop
-to use `echo` to print our commands without actually executing them.
```

Comment:
I think this is a great practical tip to add back in

Reply:
See the exercise "Doing a dry run".